SERVFAIL on fallthrough to forward

When trying to use kubernetai with the following configuration

```
.:53 {
  errors

  kubernetai cluster-a.local in-addr.arpa ip6.arpa {
    fallthrough in-addr.arpa ip6.arpa
    kubeconfig /Users/user/.kube/config arn:aws:eks:us-east-1:000000000000:cluster/a
    pods insecure
  }

  kubernetai cluster-b.local in-addr.arpa ip6.arpa {
    fallthrough in-addr.arpa ip6.arpa
    kubeconfig /Users/user/.kube/config arn:aws:eks:us-east-1:000000000000:cluster/b
    pods insecure
  }

  forward . /etc/resolv.conf {
    max_concurrent 1000
  }
}
```

and using dig to query PTR records of an IP address

```bash
dig @127.0.0.1 -x 1.1.1.1
```

it raises an error saying that the next plugin was not found

```
.:53
CoreDNS-1.8.3
darwin/amd64, go1.16.3, 4293992b-dirty
[ERROR] plugin/errors: 2 1.1.1.1.in-addr.arpa. PTR: plugin/kubernetes: no next plugin found specified
```

and answers the request with SERVFAIL

```
; <<>> DiG 9.10.6 <<>> -x 1.1.1.1 @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 57076
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;1.1.1.1.in-addr.arpa.          IN      PTR

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Thu May 06 17:11:41 -03 2021
;; MSG SIZE  rcvd: 49
```

instead of using the forward plugin configuration.

--- a/plugin/kubernetai/kubernetai.go
+++ b/plugin/kubernetai/kubernetai.go
@@ -5,9 +5,7 @@ import (
 
 	"github.com/coredns/coredns/plugin"
 	"github.com/coredns/coredns/plugin/kubernetes"
-	"github.com/coredns/coredns/plugin/pkg/fall"
 	clog "github.com/coredns/coredns/plugin/pkg/log"
-	"github.com/coredns/coredns/plugin/pkg/nonwriter"
 	"github.com/coredns/coredns/request"
 	"github.com/miekg/dns"
 )
@@ -17,9 +15,7 @@ var log = clog.NewWithPlugin("kubernetai")
 // Kubernetai handles multiple Kubernetes
 type Kubernetai struct {
 	Zones          []string
-	Next           plugin.Handler
 	Kubernetes     []*kubernetes.Kubernetes
-	Fall           []fall.F
 	autoPathSearch []string // Local search path from /etc/resolv.conf. Needed for autopath.
 	p              podHandlerItf
 }
@@ -37,47 +33,7 @@ func New(zones []string) (Kubernetai, *kubernetes.Kubernetes) {
 
 // ServeDNS routes requests to the authoritative kubernetes. It implements the plugin.Handler interface.
 func (k8i Kubernetai) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (rcode int, err error) {
-	state := request.Request{W: w, Req: r}
-	for i, k := range k8i.Kubernetes {
-		zone := plugin.Zones(k.Zones).Matches(state.Name())
-		if zone == "" {
-			continue
-		}
-
-		// If fallthrough is enabled and there are more kubernetes in the list, then we
-		// should continue to the next kubernetes in the list (not next plugin) when
-		// ServeDNS results in NXDOMAIN.
-		if i != (len(k8i.Kubernetes)-1) && k8i.Fall[i].Through(state.Name()) {
-			// Use a non-writer so we don't write NXDOMAIN to client
-			nw := nonwriter.New(w)
-
-			_, err := k.ServeDNS(ctx, nw, r)
-
-			// Return SERVFAIL if error
-			if err != nil {
-				return dns.RcodeServerFailure, err
-			}
-
-			// If NXDOMAIN, continue to next kubernetes instead of next plugin
-			if nw.Msg.Rcode == dns.RcodeNameError {
-				continue
-			}
-
-			// Otherwise write message to client
-			m := nw.Msg
-			state.SizeAndDo(m)
-			m = state.Scrub(m)
-			w.WriteMsg(m)
-
-			return m.Rcode, err
-
-		} else {
-			rcode, err = k.ServeDNS(ctx, w, r)
-		}
-
-		return rcode, err
-	}
-	return plugin.NextOrFailure(k8i.Name(), k8i.Next, ctx, w, r)
+	return k8i.Kubernetes[0].ServeDNS(ctx, w, r)
 }
 
 // AutoPath routes AutoPath requests to the authoritative kubernetes.

--- a/plugin/kubernetai/kubernetai_test.go
+++ b/plugin/kubernetai/kubernetai_test.go
@@ -166,7 +166,6 @@ func TestKubernetai_AutoPath(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			k8i := Kubernetai{
 				Zones:          tt.fields.Zones,
-				Next:           tt.fields.Next,
 				Kubernetes:     tt.fields.Kubernetes,
 				autoPathSearch: tt.fields.autoPathSearch,
 				p:              tt.fields.p,

--- a/plugin/kubernetai/setup.go
+++ b/plugin/kubernetai/setup.go
@@ -2,6 +2,7 @@ package kubernetai
 
 import (
 	"context"
+	"errors"
 
 	"github.com/coredns/caddy"
 	"github.com/coredns/coredns/core/dnsserver"
@@ -22,16 +23,22 @@ func setup(c *caddy.Controller) error {
 		return plugin.Error(Name(), err)
 	}
 
+	prev := &kubernetes.Kubernetes{}
 	for _, k := range k8i.Kubernetes {
 		err = k.InitKubeCache(context.Background())
 		if err != nil {
 			return plugin.Error(Name(), err)
 		}
 		k.RegisterKubeCache(c)
+
+		// set Next of the previous kubernetes instance to the current instance
+		prev.Next = k
+		prev = k
 	}
 
 	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
-		k8i.Next = next
+		// set Next of the last kubernetes instance to the next plugin
+		k8i.Kubernetes[len(k8i.Kubernetes)-1].Next = next
 		return k8i
 	})
 
@@ -44,23 +51,17 @@ func Parse(c *caddy.Controller) (*Kubernetai, error) {
 		autoPathSearch: searchFromResolvConf(),
 		p:              &podHandler{},
 	}
-	var err error
+
 	for c.Next() {
-		var k8s *kubernetes.Kubernetes
-		k8s, err = kubernetes.ParseStanza(c)
+		k8s, err := kubernetes.ParseStanza(c)
 		if err != nil {
 			return nil, err
 		}
 		k8i.Kubernetes = append(k8i.Kubernetes, k8s)
 	}
 
-	for i := 0; i < len(k8i.Kubernetes); i++ {
-		// Copy Fallthough settings from the Kubernetes object
-		k8i.Fall = append(k8i.Fall, k8i.Kubernetes[i].Fall)
-		// for all but last instance, disable fallthrough in the Kubernetes object
-		if i < len(k8i.Kubernetes)-1 {
-			k8i.Kubernetes[i].Fall.Zones = nil
-		}
+	if len(k8i.Kubernetes) == 0 {
+		return nil, errors.New("no kubernetes instance was parsed")
 	}
 
 	return k8i, nil

--- a/plugin/kubernetai/setup_test.go
+++ b/plugin/kubernetai/setup_test.go
@@ -1,0 +1,140 @@
+package kubernetai
+
+import (
+	"context"
+	"testing"
+
+	"github.com/miekg/dns"
+
+	"github.com/coredns/coredns/plugin/kubernetes"
+
+	"github.com/coredns/caddy"
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+)
+
+func TestSetup(t *testing.T) {
+	tests := []struct {
+		input     string
+		instances int
+		hasNext   bool
+	}{
+		{
+			input: `
+				kubernetai cluster.local {
+				  endpoint http://192.168.99.100
+				}
+			`,
+			instances: 1,
+			hasNext:   false,
+		},
+		{
+			input: `
+				kubernetai cluster.local {
+				  endpoint http://192.168.99.100
+				}
+				kubernetai assemblage.local {
+				  endpoint http://192.168.99.101
+				}
+			`,
+			instances: 2,
+			hasNext:   false,
+		}, {
+			input: `
+				kubernetai cluster.local {
+				  endpoint http://192.168.99.100
+				}
+				kubernetai assemblage.local {
+				  endpoint http://192.168.99.101
+				}
+				kubernetai conglomeration.local {
+				  endpoint http://192.168.99.102
+				}
+			`,
+			instances: 3,
+			hasNext:   false,
+		},
+		{
+			input: `
+				kubernetai cluster.local {
+				  endpoint http://192.168.99.100
+				}
+			`,
+			instances: 1,
+			hasNext:   true,
+		},
+		{
+			input: `
+				kubernetai cluster.local {
+				  endpoint http://192.168.99.100
+				}
+				kubernetai assemblage.local {
+				  endpoint http://192.168.99.101
+				}
+			`,
+			instances: 2,
+			hasNext:   true,
+		}, {
+			input: `
+				kubernetai cluster.local {
+				  endpoint http://192.168.99.100
+				}
+				kubernetai assemblage.local {
+				  endpoint http://192.168.99.101
+				}
+				kubernetai conglomeration.local {
+				  endpoint http://192.168.99.102
+				}
+			`,
+			instances: 3,
+			hasNext:   true,
+		},
+	}
+
+	for i, test := range tests {
+		var nextHandler plugin.Handler
+		if test.hasNext {
+			handlerFunc := plugin.HandlerFunc(func(_ context.Context, _ dns.ResponseWriter, _ *dns.Msg) (int, error) {
+				return 0, nil
+			})
+			nextHandler = &handlerFunc
+		}
+
+		c := caddy.NewTestController("dns", test.input)
+
+		if err := setup(c); err != nil {
+			t.Fatalf("Test %d: %v", i, err)
+		}
+
+		plugins := dnsserver.GetConfig(c).Plugin
+		if n := len(plugins); n != 1 {
+			t.Fatalf("Test %d: Expected plugin length on controller to be 1, got %d", i, n)
+		}
+
+		handler := plugins[0](nextHandler)
+
+		k8i, ok := handler.(*Kubernetai)
+		if !ok {
+			t.Fatalf("Test %d: Expected handler to be Kubernetai, got %T", i, handler)
+		}
+
+		if n := len(k8i.Kubernetes); n != test.instances {
+			t.Fatalf("Test %d: Expected kubernetes length on handler to be %d, got %d", i, test.instances, n)
+		}
+
+		prev := &kubernetes.Kubernetes{
+			Next: k8i.Kubernetes[0],
+		}
+		for j, k := range k8i.Kubernetes {
+			if prev.Next != k {
+				t.Fatalf("Test %d: Expected kubernetes instance %d to be referencing kubernetes instance %d as next, got %v", i, j-1, j, prev.Next)
+			}
+
+			prev = k
+		}
+
+		if prev.Next != nextHandler {
+			t.Fatalf("Test %d: Expected last kubernetes instance to be referencing nextHandler as next, got %v", i, prev.Next)
+		}
+	}
+}

